### DISCUSSION TITLE
Pip packages shadow Ubuntu packages ?

### DISCUSSION THREAD

**Discussion by heruji** [arkane-systems/genie#231] [Q&A] [answered]
When upgrading to 2.1 using `sudo apt upgrade`, I saw that python packages `argparse`, `contextlib2`, `psutil`, `python_hosts`, `nsenter` were installed using pip (in `/usr/local/lib/python3.8/dist-packages`). But it seems Ubuntu already has packages `python-contextlib2`, `python3-psutil`, etc, providing some of the dependencies (and they use different versions). In my default `sys.path`, `/usr/local/lib/python3.8/dist-packages` comes before `/usr/lib/python3/dist-packages`. My fear is that those pip packages shadow Ubuntu packages, which may cause issues. Is it possible to use Ubuntu packages to manage the dependency ?

**Comment by cerebrate** [comment, marked as answer]
Unfortunately, there aren't Debian/Ubuntu packages for the _nsenter_ and _python_hosts_ dependencies, and those are the ones which drag the others in.

That said, I acknowledge the potential problem here, and I'm pleased to be able to say that there's a different fix for this, internalizing the dependencies, which will ship in 2.2. (You can already see this in place on the dev-2.2 branch, FYI.)
